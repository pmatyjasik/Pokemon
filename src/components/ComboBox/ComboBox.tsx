"use client";

import { Autocomplete, AutocompleteProps, CircularProgress, Popper, styled, useTheme } from "@mui/material";
import { Input } from "@/components/Input/Input";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import { forwardRef, useMemo, Ref } from "react";

const StyledPopper = styled(Popper)(({ theme }) => ({
    "& .MuiAutocomplete-paper": {
        border: theme.palette.grey[400],
        boxShadow: theme.shadows[0],
    },
    "& .MuiAutocomplete-noOptions": {
        fontSize: 14,
    },
}));

const StyledListbox = styled("ul")(({ theme }) => ({
    fontSize: 14,
    height: theme.spacing(50),
    "& .MuiAutocomplete-option": {
        cursor: "pointer",
        "&:hover": {
            backgroundColor: `${theme.palette.primary.light} !important`,
        },
        "&[aria-selected='true']": {
            backgroundColor: `${theme.palette.background.default} !important`,
            color: theme.palette.primary.main,
        },
    },
}));

export interface ComboBoxProps<T> extends Omit<AutocompleteProps<T, false, true, false>, "renderInput"> {
    placeholder: string;
}

export const ComboBox = forwardRef(<T,>(props: ComboBoxProps<T>, ref: Ref<HTMLDivElement>) => {
    const { placeholder, ...autocompleteProps } = props;
    const { palette, spacing } = useTheme();

    const popupIcon = useMemo(() => {
        const color = { color: palette.grey[100] };

        return props.loading ? (
            <CircularProgress data-testid='loader' size={spacing(4)} sx={color} />
        ) : (
            <ExpandMoreIcon sx={color} />
        );
    }, [props.loading, palette.grey, spacing]);

    return (
        <Autocomplete
            {...autocompleteProps}
            ref={ref}
            disableClearable
            PopperComponent={StyledPopper}
            ListboxComponent={StyledListbox}
            popupIcon={popupIcon}
            renderInput={(params) => <Input {...params} placeholder={placeholder} />}
        />
    );
}) as (<T>(props: ComboBoxProps<T> & { ref?: Ref<HTMLDivElement> }) => JSX.Element) & {
    displayName?: string;
};

ComboBox.displayName = "ComboBox";
