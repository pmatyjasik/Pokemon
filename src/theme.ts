"use client";

import { createTheme, Shadows } from "@mui/material/styles";
import { ThemeOptions } from "@mui/material/styles";
import { NextFont } from "next/dist/compiled/@next/font";
import localFont from "next/font/local";

const IbmVga: NextFont = localFont({
    src: "./fonts/IBM_VGA.woff",
    display: "swap",
    style: "normal",
    weight: "400",
});

const themeOptions: ThemeOptions = {
    palette: {
        mode: "light",
        background: {
            default: "rgba(255, 255, 255, 1)",
        },
        primary: {
            main: "rgba(151, 71, 255, 1)",
            dark: "rgba(113, 53, 191, 1)",
            light: "rgba(151, 71, 255, 0.25)",
        },
        grey: {
            100: "rgba(42, 42, 42, 1)",
            200: "rgba(127, 127, 127, 1)",
            300: "rgba(228, 228, 228, 1)",
            400: "rgba(238, 238, 238, 1)",
        },
        error: {
            main: "rgba(255, 78, 78, 1)",
        },
        action: {
            disabled: "rgba(0, 0, 0, 0.2)",
        },
    },
    spacing: 4,
    typography: {
        fontFamily: IbmVga.style.fontFamily,
        fontWeightMedium: 400,
        fontSize: 16,
        h6: {
            fontSize: "12px",
        },
        caption: {
            fontSize: "10px",
        },
    },
    shape: {
        borderRadius: 2,
    },
    shadows: [
        "none",
        "0px 0px 0px 4px rgba(151, 71, 255, 0.25)",
        "0px 4px 10px 2px rgba(0, 0, 0, 0.1)",
        ...Array(20).fill("none"),
    ] as Shadows,
};

export const theme = createTheme(themeOptions);
