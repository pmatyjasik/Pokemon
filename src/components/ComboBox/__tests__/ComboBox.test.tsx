import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { ThemeProvider } from "@mui/material/styles";
import { theme } from "@/theme";
import { ComboBox } from "@/components/ComboBox/ComboBox";

const options = ["Option 1", "Option 2", "Option 3"];

const renderComponent = (props = {}) => {
    return render(
        <ThemeProvider theme={theme}>
            <ComboBox options={options} placeholder='Select an option' {...props} />
        </ThemeProvider>
    );
};

describe("ComboBox", () => {
    it("Matches DOM Snapshot", () => {
        const { asFragment } = renderComponent();

        expect(asFragment()).toMatchSnapshot();
    });

    it("should renders with placeholder text", () => {
        renderComponent();

        expect(screen.getByPlaceholderText("Select an option")).toBeInTheDocument();
    });

    it("should displays loading indicator when loading prop is true", () => {
        renderComponent({ loading: true });

        expect(screen.getByTestId("loader")).toBeInTheDocument();
    });

    it("should not display clear button", () => {
        renderComponent();

        fireEvent.click(screen.getByRole("combobox"));
        fireEvent.change(screen.getByRole("combobox"), { target: { value: "Option 1" } });

        expect(screen.queryByLabelText("Clear")).not.toBeInTheDocument();
    });
});
