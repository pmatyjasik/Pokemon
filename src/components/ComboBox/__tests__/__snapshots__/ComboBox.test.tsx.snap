// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboBox Matches DOM Snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-ka7ti6-MuiAutocomplete-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-iaqcp-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-6m1092-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          placeholder="Select an option"
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment css-p1olib-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-1n96gr0-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-6xpt3h-MuiSvgIcon-root"
              data-testid="ExpandMoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-ihdtdm"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
